testnet/stacks-node/src/config.rs

```diff
@@ -494,11 +495,44 @@ impl Config {
                     rbf_fee_increment: burnchain
                         .rbf_fee_increment
                         .unwrap_or(default_burnchain_config.rbf_fee_increment),
-                    epochs: match burnchain.epochs {
-                        Some(epochs) => Some(epochs),
-                        None => default_burnchain_config.epochs,
-                    },
+                    epochs: default_burnchain_config.epochs,
+                };
+
+                if let Some(ref conf_epochs) = burnchain.epochs {
+                    let mut epochs = match result.get_bitcoin_network().1 {
+                        BitcoinNetworkType::Mainnet => {
+                            panic!("Cannot configure epochs in mainnet mode");
+                        }
+                        BitcoinNetworkType::Testnet => stacks::core::STACKS_EPOCHS_TESTNET.to_vec(),
+                        BitcoinNetworkType::Regtest => stacks::core::STACKS_EPOCHS_REGTEST.to_vec(),
+                    };
+                    if result.mode == "mocknet" {
+                        for epoch in epochs.iter_mut() {
+                            epoch.block_limit = ExecutionCost::max_value();
+                        }
+                    }
+                    for configured_epoch in conf_epochs.iter() {
+                        let mut epoch = epochs
+                            .iter_mut()
+                            .find(|x| x.epoch_id as u32 == configured_epoch.epoch_id)
+                            .expect(&format!(
+                                "Failed to find Epoch ID in defined epochs: {} not found",
+                                configured_epoch.epoch_id
+                            ));
+                        epoch.start_height = configured_epoch
+                            .start_height
+                            .try_into()
+                            .expect("Epoch start height must be a positive integer");
+                        epoch.end_height = configured_epoch
+                            .end_height
+                            .try_into()
+                            .expect("Epoch end height must be a positive integer");
+                    }
+
+                    result.epochs = Some(epochs);
                 }
+
+                result
```

Comment:
Not sure how feasible this is, but in the spirit of @igorsyl's earlier work on doing config file validation earlier rather than later, would it be possible to verify here that the epochs are well-formed -- i.e. no block ranges overlap, and there are no gaps?

Reply:
Hmm, maybe it'll be somewhat easy, I'll see what I can do.

Reply:
The simplest way I can think of would be to do the following:

* Accept epochs by name, not value, so any typos or non-existent epochs would be quickly caught
* Accept only start heights, not end heights.  The end height for epoch _E_ is the block before the start height for epoch _E+1_.  Then, the user can't possibly create an overlapping range.
* Require that the start height for epoch _E_ is always less than all other epochs _E+k_.  If the user doesn't do this, then the error to report is intuitive: a later epoch cannot start before an earlier epoch.

Reply:
Updated in 32d10fc92a31504e59dd2c552fc85166c57df90c -- the new format takes an `epoch_name` and a `start_height`, and it checks that epochs are in the correct order, that epoch names are as expected, and that the configured epochs are a prefix of `1.0, 2.0, 2.05, 2.1` (so you cannot specify `1.0, 2.05, 2.1`, skipping `2.0`)